sensors/healthd-0.6.5/healthdc.8
.\"-
.\" Copyright (c) 1999-2000 James E. Housley <jim@thehousleys.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	$Id: healthdc.8,v 1.1 2001-12-05 18:45:09 kwebb Exp $
.\"
.\" manual page [] for healthd 0.6.1
.Dd June 12, 2000
.Dt HEALTHDC 8
.Os FreeBSD
.Sh NAME
.Nm healthdc
.Nd
Motherboard 
.Dq health
monitor
.Sh SYNOPSIS
.Nm
.Op Fl H|h
.Op Fl P Ar port
.Op Fl V
.Op Fl d
.Op Fl t|v
.Op Ar host ...

.Sh DESCRIPTION
This program provides a facility to read the current values from
local and remote 
.Xr healthd(8)
programs.  This allows monitoring of all networked computers for
critical hardware failures.  
.Nm Healthdc
can be run from other programs or scripts to automatically log
the current values and watch for trends.

.Sh OPTIONS
The following command line options are available.
.Bl -tag -width Fl

.It Fl H
HTML Output.  This will produce a full HTML page with a table for each host
specified on the command line.

.It Fl P Ar port
Use a different port then the default 1281.  This is a different port
than previous versions.  The older versions used 9669, however 1281 is
the official port number.

.It Fl V
Display the current version and exit.

.It Fl d
Display extra debugging info.  Not implemented yet.

.It Fl h
HTML output, without headers.  This functions in a similar fashion to 
.Ar -H
except the output is missing the <HTML>, <HEAD>, and <BODY> start and end
tags.  This is intended to be used within another HTML page.

.It Fl t
Display the output in a tab delimited line.  The order of the output is
hostname, temp1, temp2, temp3, fan1, fan2, fan3, Vcore1, Vcore2, 3volt,
+5Volt, +12Volt, -12Volt and -5Volt.

.It Fl v
This displays a comma delimited list with all stats about each of the
monitored parameters.  This includes, in addition to the current value, 
whether it is active, the number of out-of-bounds readings, if a failure
has been logged and if a warning has been logged.  The order of the output is
hostname, temp1, temp2, temp3, fan1, fan2, fan3, Vcore1, Vcore2, 3volt,
+5Volt, +12Volt, -12Volt and -5Volt.

.It host ...
The host, or hosts, to get the current readings from.  If not specified localhost is used.


doc/ChangeLog
stoller     2003/11/17 17:30:37 MST

  Modified files:
    db                   libdb.pm.in 
    sql                  database-fill.sql 
    tbsetup              assign_wrapper.in batch_daemon.in 
                         batchexp.in endexp.in eventsys_control.in 
                         mkexpdir.in node_update.in os_setup.in 
                         startexp.in swapexp.in tbend.in 
                         tbprerun.in tbreport.in tbswap.in 
                         vnode_setup.in 
    utils                delay_config.in 
    www                  beginexp.php3 dbdefs.php3.in 
                         delaycontrol.php3 editexp.php3 
                         modifyexp.php3 showexp.php3 
                         showexp_list.php3 showstuff.php3 
                         swapexp.php3 
  Log:
  Merge the two state machines (batchstate and state) into a single
  state machine (state). All of the stuff that was previously handled by
  using batchstate is now embedded into the one state machine. Of
  course, these mostly overlapped, so its not that much of a change,
  except that we also redid the machine, adding more states (for
  example, modify phases are now explicit. To get a picture of the
  actual state machine, on boss:
  
  		stategraph -o newstates EXPTSTATE
  		gv newstates.ps
  
  Things to note:
  
  * The "batchstate" slot of the experiments table is now used solely to
    provide a lock for batch daemon. A secondary change will be to
    change the slot name to something more appropriate, but it can
    happen anytime after this new stuff is installed.
  
  * I have left expt_locked for now, but another later change will be to remove
    expt_locked, and change it to active_busy or some such new state name in
    the state machine. I have removed most uses of expt_locked, except those
    that were necessary until there is a new state to replace it.
  
  * These new changes are an implementation of the new state machine,
    but I have not done anything fancy. Most of the code is the same as
    it was before.
  
  * I suspect that there are races with the batch daemon now, but they
    are going to be rare, and the end result is probably that a
    cancelation is delayed a little bit.
  
  Revision  Changes    Path
  1.138     +30 -77    testbed/db/libdb.pm.in
  1.35      +26 -31    testbed/sql/database-fill.sql
  1.138     +5 -5      testbed/tbsetup/assign_wrapper.in
  1.41      +76 -95    testbed/tbsetup/batch_daemon.in
  1.41      +19 -41    testbed/tbsetup/batchexp.in
  1.29      +101 -57   testbed/tbsetup/endexp.in
  1.12      +1 -0      testbed/tbsetup/eventsys_control.in
  2.5       +8 -7      testbed/tbsetup/mkexpdir.in
  1.13      +4 -22     testbed/tbsetup/node_update.in
  1.86      +2 -0      testbed/tbsetup/os_setup.in
  1.59      +45 -39    testbed/tbsetup/startexp.in
  1.45      +220 -182  testbed/tbsetup/swapexp.in
  1.34      +6 -23     testbed/tbsetup/tbend.in
  1.42      +21 -40    testbed/tbsetup/tbprerun.in
  1.46      +5 -11     testbed/tbsetup/tbreport.in
  1.28      +93 -226   testbed/tbsetup/tbswap.in
  1.35      +2 -1      testbed/tbsetup/vnode_setup.in
  1.12      +6 -11     testbed/utils/delay_config.in
  1.41      +6 -1      testbed/www/beginexp.php3
  1.65      +1 -28     testbed/www/dbdefs.php3.in
  1.7       +7 -7      testbed/www/delaycontrol.php3
  1.5       +4 -4      testbed/www/editexp.php3
  1.11      +7 -1      testbed/www/modifyexp.php3
  1.75      +14 -15    testbed/www/showexp.php3
  1.80      +2 -2      testbed/www/showexp_list.php3
  1.121     +51 -63    testbed/www/showstuff.php3
  1.21      +3 -3      testbed/www/swapexp.php3

stoller     2003/11/17 15:35:33 MST

  Modified files:
    www                  tbauth.php3 toggle.php 
  Log:
  Add web login attack detection/prevention. Two changes:
  
  * Add slots to users table to track number of failures in the last N
    seconds. If a threshold is passed (currently 4 failures in the last
    minute), the web login is disabled. Note that I do not disable the
    ops shell login at this time. Aging is passive; the values are cleared
    when login is successful, or when more then one minute has passed
    since the last failure. In other words, a burst of failures will
    disable the login, but failures over time are okay.
  
  * Add login_failures table to do exactly the same as above, except it
    is on an IP basis (REMOTE_ADDR in the server). Currently the
    threshold is 8 failures in the last two minutes, at which time all
    logins from that IP are disabled.
  
  In both cases email is sent to tbops (and the user).
  
  The constants are defined at the top of www/tbauth.in, rather then as
  site variables, to avoid pounding the DB when an attack is being
  launched.
  
  To clear a user freeze, go to the user profile page and use the
  "toggle" near the bottom.
  
  To clear an IP freeze: delete from login_failures were IP='1.1.1.1'
  
  Revision  Changes    Path
  1.40      +144 -17   testbed/www/tbauth.php3
  1.13      +18 -3     testbed/www/toggle.php

	
stoller     2003/11/05 10:53:00 MST

  Modified files:
    db                   libdb.pm.in xmlconvert.in 
    tbsetup              batchexp.in eventsys_control.in 
    tbsetup/ns2ir        program.tcl sim.tcl.in 
  Log:
  Frontend and parser portion of two event system changes:
  
  * Generate a shared secret key for the event system. This key is
    stored into the DB, and passed to the node via tmcd. It is also
    stashed into a file in the experiment directory (can be accessed
    only by the project/group members). The key is used to attach a
    HMAC (hashed message authentication) to each event, which is checked

sensors/healthd-0.6.5/healthdc.c
/*-
 * Copyright (c) 1999-2000 James E. Housley <jim@thehousleys.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: healthdc.c,v 1.1 2001-12-05 18:45:09 kwebb Exp $
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <netdb.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <err.h>
#include <errno.h>

#if !defined(HAVE_GETADDRINFO) || !defined(HAVE_GETNAMEINFO)
#include "addrinfo.h"
#endif

#include "VERSION.h"

int main(int argc, char *argv[], char *envp[]);
void PrintNormalOutput(int , char *);
void PrintHTTPOutput(int , char *);
int OpenConnection(char *, char *, int);
int GetProtocolVersion(int sock);
int GetDaemonVersion(int sock);
void ParseLine(char *);
int hex2dec(char *ptr);
void BailOut(int code);

float VerProto = 1.0;
int VerDaemon_major = 0;
int VerDaemon_minor = 6;
int VerDaemon_sub = 0;

int tabout = 0;
int verbose = 0;
int debug = 0;
int html = 0;
int html_body = 0;

int CONTENT_LENGTH = 0;
int POST=1;

#define PrintMAX 10240
char InputLine[PrintMAX];

#define MAX_TARGETS 255
char *Target[MAX_TARGETS+1];
int TargetPtr;

char *query_string = NULL;

int
main(int argc, char *argv[], char *envp[]) {
  char ch;
  int x, y, z;
  int len;
  int sock;
  struct servent *sp;
  unsigned long port;
  char HostNameBuffer[MAXHOSTNAMELEN];
  char portstr[16];

  for (sock=0; sock<=MAX_TARGETS; sock++) {
    Target[sock] = NULL;
  }
  TargetPtr = 0;

  if ((sp = getservbyname("healthd", "tcp")) == NULL) {
    port = 1281;
    strcpy(portstr, "1281");
  } else {
    port = ntohs(sp->s_port);
    sprintf(portstr, "%ld", port);
  }
  while((ch = getopt(argc, argv, "HPVdhtv")) != -1) {
    switch(ch){
    case 'H':
      html = 1;
      POST = 0;
      break;

    case 'P':
      port = atoi(optarg);
      break;

    case 'V':
      fprintf(stderr, "Version %s\n", hdVERSION);
      exit(0);
      break;

    case 'd':
      debug = 1;
      break;

    case 'h':
      html = 1;
      html_body = 1;
      POST = 0;
      break;

    case 't':
      tabout = 1;
      break;

    case 'v':
      verbose = 1;
      break;

    default:
      fprintf(stderr, "Usage: %s [-H|h] [-P port] [-V] [-d] [-t|v] [hostname ...]\n", argv[0]);
      exit(1);
    }
  }

  while (optind < argc) {
    if ((Target[TargetPtr] = (char *)malloc(1+strlen(argv[optind]))) == NULL) {
      for (sock=0; sock<TargetPtr; sock++) {
	free(Target[TargetPtr]);
      }
      exit(1);
    }
    strcpy(Target[TargetPtr], argv[optind]);
    TargetPtr++;
    optind++;
  }
  sock = 0;
  while (envp[sock] != NULL) {
    ParseLine(envp[sock]);
    sock++;
  }
  if (html) {
    if (html_body == 0) {
      printf("Content-type: text/html\n\n<HTML><TITLE>healthd</TITLE><BODY>\n");
      x = 0;
      while (envp[x] != NULL) {
	x++;
      }
    }
    if (POST) {
      /*
       * The POST (origional) method was used
       */
      y = 0;
      len = 0;
      do {
	z = fgetc(stdin);
	len++;
	InputLine[y++] = z;
	if ((z == '&') || (z == EOF) || (z == '\n') || 
	    (z == '\r') || (z == ' ')) {
	  InputLine[y-1] = '\0';
	  ParseLine(InputLine);
	  y = 0;
	}
	if (CONTENT_LENGTH != 0) {
	  if (len == CONTENT_LENGTH) {
	    InputLine[y] = '\0';
	    ParseLine(InputLine);
	    break;
	  }
	}
	if (y > (PrintMAX - 4)) {
	  printf("<BR><CENTER><H1>ERROR!!!</H1><BR><BR><H2>INPUT TOO LARGE!!!!<BR>\n");
	  y = 0;
	}
      } while ((z != EOF) && (z != '\n') && (z != '\r') && (z != ' '));
    } else {
      if (query_string != NULL) {
	/*
	 * The GET method was used
	 */
	char *ptr;

	y = 0;
	len = 0;
	ptr = query_string;
	ptr += sizeof("QUERY_STRING=") - 1;
	do {
	  z = *ptr++;
	  len++;
	  InputLine[y++] = z;
	  if ((z == '&') || (z == EOF) || (z == '\n') || 
	      (z == '\r') || (z == ' ')) {
	    InputLine[y-1] = '\0';
	    ParseLine(InputLine);
	    y = 0;
	  }
	  if (*ptr == '\0') {
	    InputLine[y] = '\0';
	    ParseLine(InputLine);
	    break;
	  }
	  if (y > (PrintMAX - 4)) {
	    printf("<BR><CENTER><H1>ERROR!!!</H1><BR><BR><H2>INPUT TOO LARGE!!!!<BR>\n");
	    y = 0;
	  }
	} while ((z != '\0') && (z != '\n') && (z != '\r') && (z != ' '));
      }
    }
  }
  if (TargetPtr == 0) {
    if ((Target[TargetPtr] = (char *)malloc(1+strlen("localhost"))) == NULL) {
      for (sock=0; sock<TargetPtr; sock++) {
	free(Target[TargetPtr]);
      }
      exit(1);
    }
    strcpy(Target[TargetPtr], "localhost");
    TargetPtr++;
  }
  for (x=0; x<TargetPtr; x++) {
    if (*Target[x] == '[' && strrchr(Target[x], ']') != NULL) { 
      /* IPv6 address in [] */
      strncpy(HostNameBuffer, Target[x] + 1, strlen(Target[x]) - 2);
      HostNameBuffer[strlen(Target[x]) - 2] = '\0';
    } else {
      strncpy(HostNameBuffer, Target[x], strlen(Target[x]));
      HostNameBuffer[strlen(Target[x])] = '\0';
    }
    if ((sock = OpenConnection(HostNameBuffer, portstr, sizeof(HostNameBuffer))) == 0) {
      continue;
    }

    if (GetProtocolVersion(sock)) {
      close(sock);
      continue;
    }
    if (GetDaemonVersion(sock)) {
      close(sock);
      continue;
    }

    if (html) {
      PrintHTTPOutput(sock, HostNameBuffer);
    } else {
      PrintNormalOutput(sock, HostNameBuffer);
    }
  }

  for (sock=0; sock<=MAX_TARGETS; sock++) {
    if (Target[sock] != NULL) {
      free(Target[sock]);
    }
  }
  if (query_string != NULL) {
    free(query_string);
  }

  if ((html) && (html_body == 0)) {
    printf("</BODY></HTML>\n");
  }
  exit(0);
}

int 
OpenConnection(char *HostName, char *portstr, int size) {
  char *hostptr;
  struct addrinfo hints, *res, *res0;
  const char *cause = NULL;
  int error;
  int sock;

  hostptr = HostName;
  memset(&hints, 0, sizeof(hints));
  hints.ai_flags = AI_CANONNAME;
  hints.ai_family = PF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;
  hints.ai_protocol = 0;
  error = getaddrinfo(hostptr, portstr, &hints, &res0);
  if (error) {
    warnx("%s: %s", hostptr, gai_strerror(error));
    if (error == EAI_SYSTEM)
      warnx("%s: %s", hostptr, strerror(errno));
    return (0);
  }

  res = res0;
  if (res->ai_canonname) {
    (void)strncpy(HostName, res->ai_canonname, size);
  }
  sock = -1;
  for (res = res0; res; res = res->ai_next) {
    sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
    if (sock < 0) {
      cause = "socket";
      continue;
    }
    if (connect(sock, res->ai_addr, res->ai_addrlen) < 0) {
      cause = "connect";
      close(sock);
      sock = -1;
      continue;
    }

    /* We made a connection */
    break;
  }
  if (sock < 0) {
    err(1, cause);
  }
  freeaddrinfo(res0);
  return(sock);
}

void
PrintNormalOutput(int sock, char *HostNameBuffer) {
  int x;
  int rval;
  char query[16];
  char seprt[8];
  char buf[1024];

  if (verbose) {
    sprintf(query, "GTV");
    sprintf(seprt, ", ");
  } else {
    sprintf(query, "GET");
    sprintf(seprt, "\t");
  }

  printf("%s%s", HostNameBuffer, seprt);
  for (x=1; x<=3; x++) {
    sprintf(buf, "%s T%d", query, x);
    if (write(sock, buf, strlen(buf)) < 0) {
      perror("sending stream message");
    }
    bzero(buf, sizeof(buf));
    if ((rval = read(sock, buf, 1024)) < 0) {
      perror("reading stream message");
    }
    if (rval == 0) {
      printf("Ending connections\n");
      goto lost_connection;
    } else {
      printf("%s%s", buf, seprt);
    }
  }
  for (x=1; x<=3; x++) {
    sprintf(buf, "%s S%d", query, x);
    if (write(sock, buf, strlen(buf)) < 0) {
      perror("sending stream message");
    }
    bzero(buf, sizeof(buf));
    if ((rval = read(sock, buf, 1024)) < 0) {
      perror("reading stream message");
    }
    if (rval == 0) {
      printf("Ending connections\n");
      goto lost_connection;
    } else {
      printf("%s%s", buf, seprt);
    }
  }
  for (x=1; x<=7; x++) {
    sprintf(buf, "%s V%d", query, x);
    if (write(sock, buf, strlen(buf)) < 0) {
      perror("sending stream message");
    }
    bzero(buf, sizeof(buf));
    if ((rval = read(sock, buf, 1024)) < 0) {
      perror("reading stream message");
    }
    if (rval == 0) {
      printf("Ending connections\n");
      goto lost_connection;
    } else {
      printf("%s%s", buf, seprt);
    }
  }
lost_connection:
  close(sock);
  printf("\n");
}

void
PrintHTTPOutput(int sock, char *HostNameBuffer) {
  int x;
  int rval;
  char buf[1024];
  int active;
  int Ival;
  float Fval;

  printf("<CENTER><H1>%s</H1></CENTER>", HostNameBuffer);
  printf("<CENTER><TABLE BORDER=2>\n");
  if (VerProto < 2.0) {
    for (x=1; x<=3; x++) {
      sprintf(buf, "GTV T%d", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      } else {
	sscanf(buf, "%f|%d|", &Fval, &active);
	if (active) {
	  printf("<TR><TD>Temp #%d</TD><TD>%4.1f</TD></TR>\n", x, Fval);
	}
      }
    }
    for (x=1; x<=3; x++) {
      sprintf(buf, "GTV S%d", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      } else {
	sscanf(buf, "%d|%d|", &Ival, &active);
	if (active) {
	  printf("<TR><TD>Fan Speed #%d</TD><TD>%d</TD></TR>\n", x, Ival);
	}
      }
    }
    for (x=1; x<=7; x++) {
      sprintf(buf, "GTV V%d", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      } else {
	sscanf(buf, "%f|%d|", &Fval, &active);
	if (active) {
	  printf("<TR><TD>Voltage #%d</TD><TD>%4.2f</TD></TR>\n", x, Fval);
	}
      }
    }
    printf("</TABLE></CENTER>\n");
  } else {
    /*
     * Version 2.0 or newer
     */
    for (x=1; x<=3; x++) {
      sprintf(buf, "CFG T%d_ACT", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      }
      if (strncasecmp(buf, "yes", strlen("yes")) == 0) {
	sprintf(buf, "CFG T%d_LBL", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	  printf("<TR><TD ALIGN=RIGHT>%s</TD>", buf);
	}
	sprintf(buf, "GET T%d", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	    printf("<TD>%s</TD></TR>\n", buf);
	}
      }
    }
    for (x=1; x<=3; x++) {
      sprintf(buf, "CFG F%d_ACT", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      }
      if (strncasecmp(buf, "yes", strlen("yes")) == 0) {
	sprintf(buf, "CFG F%d_LBL", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	  printf("<TR><TD ALIGN=RIGHT>%s</TD>", buf);
	}
	sprintf(buf, "GET S%d", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	    printf("<TD>%s</TD></TR>\n", buf);
	}
      }
    }
    for (x=1; x<=7; x++) {
      sprintf(buf, "CFG V%d_ACT", x);
      if (write(sock, buf, strlen(buf)) < 0) {
	perror("sending stream message");
      }
      bzero(buf, sizeof(buf));
      if ((rval = read(sock, buf, 1024)) < 0) {
	perror("reading stream message");
      }
      if (rval == 0) {
	printf("Ending connections\n");
	goto lost_connection;
      }
      if (strncasecmp(buf, "yes", strlen("yes")) == 0) {
	sprintf(buf, "CFG V%d_LBL", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	  printf("<TR><TD ALIGN=RIGHT>%s</TD>", buf);
	}
	sprintf(buf, "GET V%d", x);
	if (write(sock, buf, strlen(buf)) < 0) {
	  perror("sending stream message");
	}
	bzero(buf, sizeof(buf));
	if ((rval = read(sock, buf, 1024)) < 0) {
	  perror("reading stream message");
	}
	if (rval == 0) {
	  printf("Ending connections\n");
	  goto lost_connection;
	} else {
	    printf("<TD>%s</TD></TR>\n", buf);
	}
      }
    }
    printf("</TABLE></CENTER>\n");
  }
lost_connection:
  close(sock);
  printf("\n");
}

int 
GetProtocolVersion(int sock) {
  char buf[1024];
  int rval;
  float value;

  VerProto = 1.0;
  sprintf(buf, "VER P");
  if (write(sock, buf, strlen(buf)) < 0) {
    perror("sending stream message");
  }
  bzero(buf, sizeof(buf));
  if ((rval = read(sock, buf, 1024)) < 0) {
    perror("reading stream message");
  }
  if (rval == 0) {
    printf("Ending connections\n");
    return 1;
  } else {
    if (strncmp(buf, "ERROR", 5) != 0) {
      if (sscanf(buf, "%f", &value) == 1) {
	VerProto = value;
      }
    }
  }
  return 0;
}

int 
GetDaemonVersion(int sock) {
  char buf[1024];
  int rval;
  int major, minor, sub;

  VerDaemon_major = 0;
  VerDaemon_minor = 6;
  VerDaemon_sub = 0;
  sprintf(buf, "VER d");
  if (write(sock, buf, strlen(buf)) < 0) {
    perror("sending stream message");
  }
  bzero(buf, sizeof(buf));
  if ((rval = read(sock, buf, 1024)) < 0) {
    perror("reading stream message");
  }
  if (rval == 0) {
    printf("Ending connections\n");
    return 1;
  } else {
    if (strncmp(buf, "ERROR", 5) != 0) {
      if (sscanf(buf, "%d.%d.%d", &major, &minor, &sub) == 3) {
	VerDaemon_major = major;
	VerDaemon_minor = minor;
	VerDaemon_sub = sub;
      }
    }
  }
  return 0;
}

int 
hex2dec(char *ptr) {
  int rtn;

  if (('0' <= ptr[0]) && (ptr[0] <= '9'))
    rtn = ptr[0] -'0';
  else if (('A' <= ptr[0]) && (ptr[0] <= 'F'))
    rtn = ptr[0] -'A' + 10;
  else if (('a' <= ptr[0]) && (ptr[0] <= 'f'))
    rtn = ptr[0] -'a' + 10;
  else 
    rtn = 0;
  rtn *= 16;
  if (('0' <= ptr[1]) && (ptr[1] <= '9'))
    rtn += ptr[1] -'0';
  else if (('A' <= ptr[1]) && (ptr[1] <= 'F'))
    rtn += ptr[1] -'A' + 10;
  else if (('a' <= ptr[1]) && (ptr[1] <= 'f'))
    rtn += ptr[1] -'a' + 10;
  else 
    rtn += 0;
  return rtn;
}

void 
BailOut(int code) {
  if (html) {
    printf("<BR><CENTER><H1>ERROR!!!</H1><BR><BR><H2>");
  }
  printf("Internal Error #%d\n", code);
  if (html) {
    printf("</HTML>\n\n");
  }
  exit(code);
}

void 
ParseLine(char *line) {
  int x, y, len;

  if (strncmp("QUERY_STRING=", line, strlen("QUERY_STRING=")) == 0) {
    html = 1;
    if (strlen(line) != strlen("QUERY_STRING=")) {
      if ((query_string = (char *)malloc(1+strlen(line))) != NULL) {
	strcpy(query_string, line);
      } else {
	BailOut(1007);
      }
    }
  }
  else if (strncmp("HTTP_REFERER=", line, strlen("HTTP_REFERER=")) == 0) {
    html = 1;
  }
  else if (strncmp("REQUEST_METHOD=", line, strlen("REQUEST_METHOD=")) == 0) {
    html = 1;
    if (strlen(line) != strlen("REQUEST_METHOD=")) {
      if (strncmp("REQUEST_METHOD=GET", line, strlen("REQUEST_METHOD=GET")) == 0) {
	POST = 0;
      }
    }
  }
  else if (strncmp("HTTP_CONTENT_LENGTH=", line, strlen("HTTP_CONTENT_LENGTH=")) == 0) {
    int length;

    html = 1;
    if (sscanf(line, "HTTP_CONTENT_LENGTH=%d", &length) == 1) {
      CONTENT_LENGTH = length;
    }
  }
  else if (strncmp("CONTENT_LENGTH=", line, strlen("CONTENT_LENGTH=")) == 0) {
    int length;

    html = 1;
    if (sscanf(line, "CONTENT_LENGTH=%d", &length) == 1) {
      CONTENT_LENGTH = length;
    }
  }
  else if (strncmp("SERVER_ADMIN=", line, strlen("SERVER_ADMIN=")) == 0) {
    html = 1;
  } else {
    /*
     * Process the rest
     */
    len = strlen(line);
    y = 0;
    for (x=0; x<=len; x++) {
      if (line[x] == '+') {
	line[y++] = ' ';
      }
      else if (line[x] == '%') {
	line[y++] = hex2dec(line+(x+1));
	x += 2;
	if ((line[y-2] == 0x0D) && (line[y-1] == 0x0A)) {
	  line[y-2] = 0x0A;
	  y -= 1;
	}
      } else {
	line[y++] = line[x];
      }
    }
    y--;
    line[y] = '\0';
    if (strncmp("HEALTHD_TARGET=", line, strlen("HEALTHD_TARGET=")) == 0) {
      if (strlen(line) != strlen("HEALTHD_TARGET=")) {
	/* Not the first, already data in user->uname */
	if ((Target[TargetPtr] = (char *)malloc(1+strlen(line))) != NULL) {
	  strcpy(Target[TargetPtr], line+sizeof("HEALTHD_TARGET"));
	  TargetPtr++;
	} else {
	  BailOut(1007);
	}
      }
    }
  }
}
